### PR TITLE
QuickCheck trace generators

### PR DESCRIPTION
**Pull request by dnadales**
Add a `QuickCheck` `HasTrace` class, and associated functions.

### CODE REVIEW

**Comment by erikd** [comment]
@dnadales Please use a proper PR title rather than just using the branch name.

**Comment by dnadales** [comment]
> @dnadales Please use a proper PR title rather than just using the branch name.

Done. Thanks for pointing this out.

**Comment by nc6** [comment]
LGTM!

File: byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+
+  envGen :: traceGenEnv -> QuickCheck.Gen (Environment sts)
+
+  sigGen
+    :: traceGenEnv
+    -> Environment sts
+    -> State sts
+    -> QuickCheck.Gen (Signal sts)
+
+  shrinkSignal :: Signal sts -> [Signal sts]
```
Comment (mhuesch) [inline byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs]: Nice.

File: byron/semantics/executable-spec/test/Control/State/Transition/Examples/CommitReveal.hs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+-- > commit (hash d0) -> commit (hash d0)
+--
+-- where it shouldn't be possible to shrink @d0@ any further.
+--
+prop_qc_OnlyValidSignals :: QC.Property
+prop_qc_OnlyValidSignals = QC.withMaxSuccess 500 -- We need to test a large
+                                                 -- number of times to make sure
+                                                 -- we get a collision in the
+                                                 -- generated data
+  $ STS.Gen.onlyValidSignalsAreGenerated @(CR ShortHash Map Data) @() 100 ()
```
Comment (mhuesch) [inline byron/semantics/executable-spec/test/Control/State/Transition/Examples/CommitReveal.hs]: 👍

File: byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+  :: forall sts traceGenEnv
+   . (HasTrace sts traceGenEnv)
+  => Word64
+  -- ^ Maximum trace length.
+  -> traceGenEnv
+  -> Environment sts
+  -> State sts
+  -> QuickCheck.Gen (Trace sts)
+traceFrom maxTraceLength traceGenEnv env st0 = do
+  choosenTraceLength <- QuickCheck.choose (0, maxTraceLength)
```
[the remark this answers is not in the document]
Comment (dnadales) [reply byron/semantics/executable-spec/src/Control/State/Transition/Trace/Generator/QuickCheck.hs]: Good catch. Thanks. I'll change this.

**Pull request merged**